### PR TITLE
feat: add gender type/sex

### LINKED ISSUE #1182: Translating genders leads to wrong gender in relationships
Thanks for Monica! I recently rediscovered it and it's really helpful.

One thing I ran across: I translated the genders to german ("man" -> "männlich", "woman" -> "weiblich") using the personalization feature in my instance of Monica. However, apparently this leads to all relationships displayed as "male". For instance, my daughters have "son" beside them in the sidebar. My wife has "father" on the side. This is due to [`getLocalizedName()`](https://github.com/monicahq/monica/blob/ae521f52266d194094863c49c534cdc51dec3057/app/RelationshipType.php#L77) testing for the strings "man" or "woman" directly.

I would propose adding an attribute on the custom genders defining which pronoun to use. This attribute should be used in `getLocalizedName()` for deciding which gender to use. I'll be happy to try to create a PR for this (even though I'm not really fluent in PHP, if someone else would like to do it, I'd rather defer).

### LINKED ISSUE #1590: Localized genders not displayed on page /settings/personalization in spite of l10n being present in settings.php
On the page https://app.monicahq.com/settings/personalization the genders and the word "contacts" are only shown in english, not in the chosen l10n (see screenshot), in spite of the localizations being present in the respective l10n files e.g. monica/resources/lang/de/settings.php; reproducible at least on the german and on the french localization.

**Screenshots**

bildschirmfoto 2018-07-22 um 23 50 00

**Which version are you using:**
 - Hosting version on https://app.monicahq.com

### LINKED ISSUE #1627: When adding new person, put man or woman as default in gender drop-down
Gender drop-down is compulsory and at he moment it defaults to nothing.
If we make it default to "man" or "woman" then this will match nearly 50% of the choices. Much better than current - 0% :).

### LINKED ISSUE #1981: Gender personnalisation and translation
For now, the gender is a list of translated gender, unique for each account.
In english, "Man","Woman", and other. So the vcard import/export can gen the gender of the contact with "Man"/"Woman" parsing.
It's bad for some reasons
- if the user change the label of one gender
- for other langages
-if the user change the langage (See #1590)

Solution:
- add the information for the Sex of the gender in another column in the table
- add the ability to set the sex of the gender in the settings
- add the ability to set the default gender for a new user (See #1627)
- add the ability to not set a gender at all

Question: what to do on vcard import if 2 genders have the same gender? Took the first one? We could also limit only 1 gender linked with 1 sex.

### LINKED ISSUE #2095: vCard import sets Gender to "vCard"
**Describe the bug**
I imported a bunch of vCard contacts (exported from Google Contacts). All of them have their gender set to "vCard".  This gender type didn't exist before the import.

**Screenshots**
screen shot 2018-11-25 at 1 55 01 pm
screen shot 2018-11-25 at 1 55 11 pm

**Which version are you using:**
 - Hosting version on https://app.monicahq.com

**Additional context**
- The Google export uses vCard format 3.0
- According to the vCard format, Gender as a property is only supported in vCard 4.0 [1]
- I spot-checked both (a) contacts that had _no_ gender defined, and (b) contacts that did have gender defined. Monica applied the gender of "vCard" to both.
- Google exports gender as part of the "NOTES" property, like so:
```
VERSION:3.0
FN:Leonardo redacted
N:redacted;Leonardo;;;
NOTE:Gender: Male
END:VCARD
```

**Ideal Behavior**:
1. Monica should _not_ create and set the gender "vCard" to all imported contacts.
2. If the gender is not specified, then Monica should import the contact with no gender specified.
3. If the gender is specified inside the NOTE property... well, personally I don't know think Monica should attempt to parse genders from notes, but I'll leave that up to the maintainers.  The important piece of behavior really is (1).

Thanks!

[1] https://en.wikipedia.org/wiki/VCard

### LINKED ISSUE #2179: Make gender optional when creating a contact
This has been requested by a lot of people and I think it makes sense. There is a lot of impact on the codebase though to make it work.

### LINKED ISSUE #2309: Family relations are not shown correctly, e.g. a female father.
**Describe the bug**
Create a person with gender "female". Assign this person as another person's "father/mother". The first contact is shown as the second person's father on the second person's detail page. It should be shown as "mother" instead. 

**Additional context**
Both persons were imported via a vCard set which had no gender information. The gender was added later. 

I understand that the automatic determination of brother/sister or father/mother cannot work for non-binary people, so maybe an auto-determination is not the best idea. Maybe just add a distinct entry in the dropdown for selecting family relations.

### PR DESCRIPTION
**Pull request by asbiin**
First step for gender selection problems
Fix #1981 
Fix #1590
Fix #2309
Fix #1182

- Add a sex type behind the gender: Male, Female, Other, Unknown, None
  This type is consistent with [vCard Gender list](https://tools.ietf.org/html/rfc6350#section-6.2.7)
  This will fix #2095 if vcard uses the right field
- Make gender optional on a contact profile, fix #2179
- Add a default gender selection on gender settings, fix #1627
    * this is the default gender selected on a new contact creation
    * this will be the gender used for relationship labelling in case the contact has no gender
    * this gender will be the first one selected on multiple relationship, like: "John Doe's ant/uncle" is the default gender is a female one.

### CODE REVIEW

**Comment by djaiss** [comment]
Interesting perspective. I think it's a good move.

File: tests/Unit/Services/Contact/Contact/CreateContactTest.php [hunk 1]
```diff
@@ -48,6 +48,39 @@ public function test_it_stores_a_contact()
         );
     }
 
+    public function test_it_stores_a_contact_without_gender()
```
Comment (djaiss) [inline tests/Unit/Services/Contact/Contact/CreateContactTest.php]: Here perhaps we could add a test asserting that the contact effectively has a gender_id set to null.

File: app/Http/Controllers/Settings/GendersController.php [hunk 2]
```diff
@@ -104,4 +129,27 @@ public function destroy(GendersRequest $request, Gender $gender)
...
+    public function updateDefault(GendersRequest $request, Gender $gender)
+    {
+        $account = auth()->user()->account;
+        $account->default_gender_id = $gender->id;
+        $account->save();
+
+        return $this->format($gender);
+    }
+
+    private function format($gender)
```
Comment (djaiss) [inline app/Http/Controllers/Settings/GendersController.php]: `format` is not a good name for this method. Also, no documentation for this function?

File: database/migrations/2019_02_09_200203_add_gender_type.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    public function up()
+    {
+        Schema::table('genders', function (Blueprint $table) {
+            $table->string('type')->after('name')->nullable();
+        });
+
+        $woman = null;
+        $man = null;
+        $other = null;
+        $woman_en = trans('app.gender_female', [], 'en');
```
Comment (djaiss) [inline database/migrations/2019_02_09_200203_add_gender_type.php]: Here again, choose either camelCase or `_`, but not a mix of both.
Therefore the variable should probably be called `womanEn` or something.

File: resources/assets/js/components/settings/Genders.vue [hunk 4]
```diff
@@ -191,13 +241,23 @@ export default {
 
   methods: {
     prepareComponent() {
-      this.getGenders();
+      axios.all([
```
[the remark this answers is not in the document]
Comment (asbiin) [reply resources/assets/js/components/settings/Genders.vue]: It uses the [Promise.all()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) API which call all of the promises passed as an iterable in parallel. We gains some ms like this to open a page.

File: database/migrations/2019_02_17_112452_default_gender.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('accounts', function (Blueprint $table) {
+            $table->unsignedInteger('default_gender_id')->after('default_time_reminder_is_sent')->nullable();
+            $table->foreign('default_gender_id')->references('id')->on('genders');
```
[the remark this answers is not in the document]
Comment (asbiin) [reply database/migrations/2019_02_17_112452_default_gender.php]: Thanks, I was not aware of this feature.

File: database/migrations/2019_02_17_112452_default_gender.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+                }
+            });
+    }
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
```
[the remark this answers is not in the document]
Comment (asbiin) [reply database/migrations/2019_02_17_112452_default_gender.php]: Maybe, but I like to add them.
It's used if you run `php artisan migrate:rollback`

File: resources/assets/js/components/settings/Genders.vue [hunk 7]
```diff
@@ -274,6 +338,18 @@ export default {
           }
         });
     },
+
+    sex(type) {
```
[the remark this answers is not in the document]
Comment (asbiin) [reply resources/assets/js/components/settings/Genders.vue]: This function is not used anymore

File: resources/assets/js/components/settings/Genders.vue [hunk 8]
```diff
@@ -274,6 +338,18 @@ export default {
...
+      var t = _.find(this.genderTypes, ['id', type]);
+      return t ? t.name : '';
+    },
+
+    mdefault(id) {
```
[the remark this answers is not in the document]
Comment (asbiin) [reply resources/assets/js/components/settings/Genders.vue]: I've renamed it

File: database/migrations/2019_02_17_112452_default_gender.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
+<?php
```
[the remark this answers is not in the document]
Comment (asbiin) [reply database/migrations/2019_02_17_112452_default_gender.php]: OK

File: app/Helpers/GendersHelper.php [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+namespace App\Helpers;
+
+class GendersHelper
+{
+    /**
+     * Return a collection of genders.
+     *
+     * @return \Illuminate\Support\Collection
+     */
+    public static function getGendersInput()
```
[the remark this answers is not in the document]
Comment (asbiin) [reply app/Helpers/GendersHelper.php]: Indeed

File: database/migrations/2019_02_20_205744_allow_gender_null.php [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('contacts', function (Blueprint $table) {
+            $table->integer('gender_id')->nullable()->change();
```
[the remark this answers is not in the document]
Comment (asbiin) [reply database/migrations/2019_02_20_205744_allow_gender_null.php]: I think there is not foreign key between contacts and gender right now.

File: tests/Unit/Services/Contact/Contact/CreateContactTest.php (hunk 1 above)
Comment (asbiin) [reply tests/Unit/Services/Contact/Contact/CreateContactTest.php]: done

File: resources/assets/js/components/settings/Genders.vue [hunk 12]
```diff
@@ -20,22 +20,48 @@
...
             {{ gender.name }}
             <span class="i">
-              ({{ gender.numberOfContacts }} contacts)
+              {{ $tc('settings.personalization_genders_list_contact_number', gender.numberOfContacts, { count:gender.numberOfContacts }) }}
             </span>
           </div>
         </div>
+        <div class="dtc">
+          <div class="pa2">
+            {{ $t('settings.personalization_genders_'+gender.type.toLowerCase()) }}
```
[the remark this answers is not in the document]
Comment (asbiin) [reply resources/assets/js/components/settings/Genders.vue]: thanks

File: app/Http/Controllers/Settings/GendersController.php (hunk 2 above)
Comment (asbiin) [reply app/Http/Controllers/Settings/GendersController.php]: I've renamed it `formatData`

File: database/migrations/2019_02_09_200203_add_gender_type.php (hunk 3 above)
Comment (asbiin) [reply database/migrations/2019_02_09_200203_add_gender_type.php]: thanks

File: resources/assets/js/components/settings/Genders.vue [hunk 13]
```diff
@@ -20,22 +20,48 @@
             {{ $t('settings.personalization_contact_field_type_table_name') }}
           </div>
         </div>
+        <div class="dtc">
+          <div class="pa2 b">
+            {{ $t('settings.personalization_contact_field_type_table_sex') }}
+          </div>
+        </div>
+        <div class="dtc">
+          <div class="pa2 b">
```
[the remark this answers is not in the document]
Comment (asbiin) [reply resources/assets/js/components/settings/Genders.vue]: I've change the UX here

File: database/migrations/2019_02_20_205744_allow_gender_null.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (djaiss) [reply database/migrations/2019_02_20_205744_allow_gender_null.php]: You are right.

File: resources/lang/en/settings.php [hunk 14]
```diff
@@ -321,13 +321,29 @@
     'personalization_genders_add' => 'Add new gender type',
     'personalization_genders_desc' => 'You can define as many genders as you need to. You need at least one gender type in your account.',
     'personalization_genders_modal_add' => 'Add gender type',
-    'personalization_genders_modal_question' => 'How should this new gender be called?',
     'personalization_genders_modal_edit' => 'Update gender type',
-    'personalization_genders_modal_edit_question' => 'How should this new gender be renamed?',
+    'personalization_genders_modal_name' => 'Name',
+    'personalization_genders_modal_name_help' => 'The name used to display the gender in a contact page.',
+    'personalization_genders_modal_sex' => 'Sex',
+    'personalization_genders_modal_sex_help' => 'Used to defines the relationships, and in VCard import/export process.',
```
Comment (djaiss) [inline resources/lang/en/settings.php]:
```suggestion
    'personalization_genders_modal_sex_help' => 'Used to define the relationships, and during the VCard import/export process.',
```

File: resources/lang/en/settings.php [hunk 15]
```diff
@@ -321,13 +321,29 @@
     'personalization_genders_add' => 'Add new gender type',
     'personalization_genders_desc' => 'You can define as many genders as you need to. You need at least one gender type in your account.',
     'personalization_genders_modal_add' => 'Add gender type',
-    'personalization_genders_modal_question' => 'How should this new gender be called?',
     'personalization_genders_modal_edit' => 'Update gender type',
-    'personalization_genders_modal_edit_question' => 'How should this new gender be renamed?',
+    'personalization_genders_modal_name' => 'Name',
+    'personalization_genders_modal_name_help' => 'The name used to display the gender in a contact page.',
```
Comment (djaiss) [inline resources/lang/en/settings.php]:
```suggestion
    'personalization_genders_modal_name_help' => 'The name used to display the gender on a contact page.',
```

File: resources/views/people/edit.blade.php [hunk 16]
```diff
@@ -98,7 +98,7 @@
             <form-select
               :options="{{ $genders }}"
               value="{{ $contact->gender_id }}"
-              :required="true"
+              :required="false"
```
Comment (djaiss) [inline resources/views/people/edit.blade.php]: Editing a gender and setting a "blank" gender now does not work. It triggers an error saying the field is mandatory.

image

File: CHANGELOG [hunk 17]
```diff
@@ -7,7 +7,7 @@ New features:
 
 Enhancements:
 
-*
+* Add a sex type behind the gender
```
Comment (djaiss) [inline CHANGELOG]: There is also an important new change: gender is no more mandatory.

File: app/Helpers/GendersHelper.php [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        $genders = auth()->user()->account->genders->map(function ($gender) {
+            return [
+                'id' => $gender->id,
+                'name' => $gender->name,
+            ];
+        });
+        $genders = CollectionHelper::sortByCollator($genders, 'name');
+        $genders->prepend(['id' => '', 'name' => '']);
```
Comment (djaiss) [inline app/Helpers/GendersHelper.php]: I don't like that very much. Instead of providing a blank field, the choice should be `No gender`. Otherwise I think the user is confused.

File: CHANGELOG (hunk 17 above)
Comment (asbiin) [reply CHANGELOG]: indeed

File: resources/views/people/edit.blade.php (hunk 16 above)
Comment (asbiin) [reply resources/views/people/edit.blade.php]: Fixed

File: app/Helpers/GendersHelper.php (hunk 18 above)
Comment (asbiin) [reply app/Helpers/GendersHelper.php]: Yes, I've changed it

**Comment by asbiin** [comment]
@djaiss I think we are good to go. For your instance, my tests on the migration indicates it will take 8min

**Comment by djaiss** [comment]
@asbiin let's go merge this beast.
